content/blog/en/dojo-middleware.md

```diff
@@ -0,0 +1,213 @@
+		validCss = valid ? css.valid : css.invalid;
+	}
+	return <div classes={[css.root, validCss]}>{children()}</div>;
+});
+
+export default PhoneValidator;
+```
+
+The `PhoneValidator` uses some middleware that returns the a `valid` property that is either `true` or `false`. It will also return the `value` of the phone number that was tested. Based on whether the phone number is valid or not, it will use some CSS for a red or green border.
+
+Notice that the `phone` property to the middleware is not passed explicitly to the middleware. By provide the `phoneNumberMiddleware` as a middleware to the `PhoneValidator` widget, the middleware will have access to the properties of the widget. Let's see what that looks like.
+
+> src/middleware/phoneNumberMiddleware.tsx
+
+```tsx
+import { create } from '@dojo/framework/core/vdom';
+
+const factory = create().properties<{ phone?: string }>();
+
+export const phoneNumberMiddleware = factory(({ properties }) => {
+	return () => {
+		// extract the `phone` value from the properties of
+		// the parent widget
+		const { phone } = properties();
+		// test the phone number
+		const valid = /^\(?(\d{3})\)?[- ]?(\d{3})[- ]?(\d{4})$/.test(phone || '');
+		return {
+			valid,
+			value: phone
+		};
+	};
+});
+
+export default phoneNumberMiddleware;
+```
+
+The middleware returns a function that will test the phone number and return whether it is valid or not.
+
+Here is what this looks like in a sample application.
+
+!(https://codesandbox.io/embed/dojo-custom-middleware-w1f7m?fontsize=14&module=%2Fsrc%2Fmiddleware%2FphoneNumberMiddleware.ts)
+
+## Geolocation Middleware
+
+You could also do some fun middleware that interacts with the DOM of you widgets. For example, there is the [`intersection`](https://dojo.io/learn/middleware/available-middleware#intersection) and [`resize`](https://dojo.io/learn/middleware/available-middleware#resize) middleware.
+
+You could use a similar pattern to grab the browsers geolocation.
+
+> src/middleware/geolocation.ts
+
+```ts
+import { create } from '@dojo/framework/core/vdom';
+import icache from '@dojo/framework/core/middleware/icache';
+
+const factory = create({ icache });
+
+type Coords = Pick<Coordinates, 'latitude' | 'longitude'>;
+
+// utility to get current geolocation
+const getGeolocation = async (): Promise<Coords> => {
+	return new Promise((resolve) => {
+		if (!('geolocation' in navigator)) {
+			resolve({ latitude: 0, longitude: 0 });
+		} else {
+			navigator.geolocation.getCurrentPosition(({ coords }) => {
+				const { latitude, longitude } = coords;
+				resolve({ latitude, longitude });
+			});
+		}
+	});
+};
+
+// default coordinates
+const defaultCoordinates = { latitude: 0, longitude: 0 };
+
+export const geolocation = factory(({ middleware: { icache } }) => {
+	return (): Coords => {
+		// get current value or default
+		const coords = icache.getOrSet('coords', defaultCoordinates);
+		if (coords.latitude === 0 && coords.longitude === 0) {
```

Comment:
Random note, what if someone is actually at 0,0... would that get reported as 0.000, 0.000 and could you differentiate that with this example (I'm being a pedant here, but it's a common challenge for people new to JS for non-geolocation examples :) )

Reply:
although the example is not meant to be a robust example in terms of a geo impl but to just show the middleware pattern, i agree we should probably just return `undefined` when no location is found/or is pending.